TDataFrame.hxx

```diff
@@ -22,828 +22,1042 @@
+      {
+         auto maxOp = std::make_shared<Internal::Operations::MaxOperation>(maxV.get(), nSlots);
+         auto maxOpLambda = [maxOp](unsigned int slot, const BranchType &v) mutable { maxOp->Exec(v, slot); };
          BranchList bl = {theBranchName};
-         using DFA = TDataFrameAction<decltype(maxOpLambda), Derived>;
-         thisFrame->Book(std::unique_ptr<DFA>(new DFA(maxOpLambda, bl, *thisFrame->fDerivedPtr)));
-         return TActionResultPtr<ActionResultType>(maxV, thisFrame->GetDataFrame());
+         using DFA_t = Internal::TDataFrameAction<decltype(maxOpLambda), Proxied>;
+         thisFrame->Book(std::make_shared<DFA_t>(maxOpLambda, bl, thisFrame->fProxiedPtr));
+         return TActionResultPtr<ActionResultType>(maxV, thisFrame->GetDataFrameChecked());
       }
    };
 
-   template<typename BranchType, typename ThisType, typename ActionResultType>
-   struct SimpleAction<BranchType, ActionResultType, EActionType::kMean, ThisType> {
-      static TActionResultPtr<ActionResultType> BuildAndBook(ThisType thisFrame, const std::string& theBranchName, std::shared_ptr<ActionResultType> meanV, unsigned nSlots) {
-         auto meanOp = std::make_shared<Operations::MeanOperation>(meanV.get(), nSlots);
-         auto meanOpLambda = [meanOp](unsigned slot, const BranchType& v) mutable { meanOp->Cumulate(v, slot);};
+   template <typename BranchType, typename ThisType, typename ActionResultType>
+   struct SimpleAction<BranchType, ActionResultType, Internal::EActionType::kMean, ThisType> {
+      static TActionResultPtr<ActionResultType> BuildAndBook(ThisType thisFrame, const std::string &theBranchName,
+                                                             std::shared_ptr<ActionResultType> meanV, unsigned int nSlots)
+      {
+         auto meanOp = std::make_shared<Internal::Operations::MeanOperation>(meanV.get(), nSlots);
+         auto meanOpLambda = [meanOp](unsigned int slot, const BranchType &v) mutable { meanOp->Exec(v, slot); };
          BranchList bl = {theBranchName};
-         using DFA = TDataFrameAction<decltype(meanOpLambda), Derived>;
-         thisFrame->Book(std::unique_ptr<DFA>(new DFA(meanOpLambda, bl, *thisFrame->fDerivedPtr)));
-         return TActionResultPtr<ActionResultType>(meanV, thisFrame->GetDataFrame());
+         using DFA_t = Internal::TDataFrameAction<decltype(meanOpLambda), Proxied>;
+         thisFrame->Book(std::make_shared<DFA_t>(meanOpLambda, bl, thisFrame->fProxiedPtr));
+         return TActionResultPtr<ActionResultType>(meanV, thisFrame->GetDataFrameChecked());
       }
    };
 
-   template<typename BranchType, EActionType ActionType, typename ActionResultType>
-   TActionResultPtr<ActionResultType> CreateAction(const std::string& theBranchName, std::shared_ptr<ActionResultType> r) {
+   template <typename BranchType, Internal::EActionType ActionType, typename ActionResultType>
+   TActionResultPtr<ActionResultType> CreateAction(const std::string & theBranchName,
```

Comment:
we should have `template<typename BranchType> CreateAction<BranchType>` that directly calls `BuildAndBook<BranchType>` and a specialization `CreateAction<GuessTheType>` that does type guessing, so we can do it only when users do not explicitly specify the branch type.